### ISSUE TITLE
[Workstream B] PR 2 of 3: Remove redundant RAG work and enforce model-aware inference budgets

### ISSUE THREAD

**Issue by zaxbysauce** [ZaxbyHub/ragappv3#511]
**Roadmap:** B2 · **Workstream B: Retrieval, inference quality and multimodal evidence**
**Sequence:** PR 2 of 3 (stable slot; dependencies govern execution)
**Rebaseline:** 2026-09-05 · **Roadmap index:** #229 (F3)

## Outcome

Retrieval and generation spend predictable work under load without dropping evidence, while embedding/reranking providers and model changes behave consistently.

## Dependencies and parallel ownership

Merge prerequisites: #510 (B1).

Owning findings: EMBED-001, RERANK-003, EMBED-002, RERANK-001, RERANK-002, RERANK-004

## Required implementation and qualification

Compute query variants lazily for the actual selected route; reuse lexical work, merge/deduplicate evidence before the final rerank, and avoid embedding variants that decomposition discards. Overlap independent Wiki/KMS/raw retrieval where context precedence permits; do not serialize optional work ahead of every raw search. Preserve per-facet recall and score semantics.
Enforce a total model-aware prompt budget across system instructions, history, retrieved context, memories, Wiki/KMS, visual observations and reserved answer tokens. Use supported tokenizer accounting with explicit conservative fallback; surface finish_reason truncation and summaries of omitted context rather than silently losing evidence. Retain unique short facts.
Repair Ollama batch endpoint/payload parity, embedding in-flight config snapshots, local reranker reload on model change, singleton no-score status, feedback weighting per document, and double normalization of TEI scores. Validate legacy enrichment response objects before using auxiliary fields. Move blocking optional Redis operations and expensive sentence dedup off the event loop with bounded work and cancellation semantics.
Coordinate admission limits with E3 instead of multiplying process-local budgets per caller. Do not change model families, increase retrieval breadth, drop OCR, or shrink quality stages merely to make timing look better.

## Additional source obligations

This slot is an explicit user-requested enhancement or a consolidated vertical implementation; all obligations below are required.

## Required tests and evidence

Count provider calls on production entry paths for standard/multi-query/decomposition/hybrid/agentic modes; paired facets must remain represented after consolidation. Full/no cache and failures; provider/model change mid-flight; singleton and multi-item rerank score types; repeated chunks from one feedback-voted document; all prompt sections near budget; non-English/code/table token accounting; token limit finish; slow Redis and long passages while another chat proceeds. No acceptance based on AST/fake call counts alone: include real provider-contract fixtures and F1 quality deltas.

## Rollout and rollback

Measure baseline and record configuration. Stage performance changes independently behind measured profiles with rollback to prior behavior; model selection is F2, migrations C1/C2. Owns ENH-CTX-001, ENH-PERF-001/002 and ENH-LEGACY-001 themes, plus chat/RAG deep performance findings.

## Execution and closure contract

PR n of 3 is a stable tracking slot, not a GitHub PR number or a requirement to serialize every workstream. Explicit dependencies govern merge order. Different workstreams can run in parallel on agreed contracts; overlapping edits must be coordinated and rebased. Re-anchor current master and active PR overlap before implementation. Evidence baseline is Astra's 0728825; planning additionally reviewed master d5a9eaa (#481/#493). The live deployed revision remains unverified.

- [ ] Every owned finding below has a fresh disposition and production caller trace; already-fixed scope closes on evidence, never unnecessary code.
- [ ] Every required behavior is wired through actual API/store/UI/worker consumers and config/docs; no silent deferral or dead implementation.
- [ ] Regression tests exercise real behavior and fail when the repair is reverted. Keep unrelated supported behavior and existing safeguards intact.
- [ ] Run the repository's current CI-compatible checks for the changed surface, plus real integration/live gates named here. Do not equate mocked method probes with end-to-end proof.
- [ ] Include rollout, migration/compatibility, rollback, operator-visible outcomes, and a scoped pending release note where required by repository conventions.
- [ ] Link the implementation PR and evidence; use Closes #this-issue only after the whole scope is satisfied. Parent F3 consumes evidence and is not a circular prerequisite.

Security and related audit work are excluded. Existing safeguards remain requirements of normal implementation; excluded backlog is preserved in Workstream X without a claim of resolution. This issue authorizes planning/tracking; it is not permission to deploy, merge dependency PRs, or alter live model settings automatically.

## Owned audit evidence and regression obligations

These are historical evidence excerpts, not assertions that unchanged line numbers or deployed behavior persist. Planning statuses override stale wording. Each entry remains an individually testable obligation within this larger PR.

### EMBED-001 — Ollama batch embedding sends the modern payload to the legacy endpoint.

**Disposition at planning:** Historical confirmed finding; re-anchor before implementation.

Multi-chunk ingestion using the supported Ollama mode fails embedding batches while single-query embedding may work. This is provider-specific, not an OpenAI or TEI failure. Evidence: `backend/app/services/embeddings.py:304-308`.

**Required repair direction:** Make Ollama single and batch endpoint/payload handling consistent: explicitly support /api/embed with model/input or issue bounded individual prompt requests to the legacy endpoint. Preserve response ordering and partial-failure behavior.

**Regression that must fail before repair:** Use an HTTP contract fixture that rejects unsupported request fields and models both legacy and modern endpoints. Require two passages to produce two ordered nonempty vectors. Reintroducing input on the legacy endpoint must fail.

**Evidence limit:** Deterministic HTTP contract probe, not a running Ollama/model integration. Official protocol corroboration: https://raw.githubusercontent.com/ollama/ollama/main/docs/api.md, Generate Embedding and Generate Embeddings sections. No security assessment.

Source at audited revision: [backend/app/services/embeddings.py:304-308](https://github.com/ZaxbyHub/ragappv3/blob/0728825ad68fff81a016be2e4fc2f97a190d0621/backend/app/services/embeddings.py#L304)
```

        # Check for explicit paths
        if "/api/embeddings" in path:
            # Already has Ollama path, use as-is
            return ("ollama", base_url)
```

### RERANK-003 — Singleton bypass reports reranking success without computing a score.

**Disposition at planning:** Historical confirmed finding; re-anchor before implementation.

A distant singleton survives the configured distance threshold and its distance can be displayed as a high relevance score. Evidence: `backend/app/services/reranking.py:144-148`.

**Required repair direction:** Score singleton inputs or return success=False when scoring is bypassed.

**Regression that must fail before repair:** A singleton without a computed rerank score must retain distance filtering and distance score semantics.

**Evidence limit:** Actual service/filter runtime plus exact main-engine wiring; no live model required.

Source at audited revision: [backend/app/services/reranking.py:144-148](https://github.com/ZaxbyHub/ragappv3/blob/0728825ad68fff81a016be2e4fc2f97a190d0621/backend/app/services/reranking.py#L144)
```
        """
        n = top_n or self.top_n
        if not chunks:
            return ([], True)
        if len(chunks) <= 1:
```

### EMBED-002 — Changing embedding provider breaks requests already in flight.

**Disposition at planning:** Historical confirmed finding; re-anchor before implementation.

An embedding request overlapping a provider-format change fails unnecessarily. New requests after the transition are unaffected by this specific race. Evidence: `backend/app/services/embeddings.py:132-136`.

**Required repair direction:** Capture URL, mode, model and prefixes once for each request and use that same snapshot for payload construction, response parsing and metrics.

**Regression that must fail before repair:** Hold an OpenAI response, switch live configuration to TEI, release the response and require the in-flight request to return its valid vector; require a subsequent request to use TEI.

**Evidence limit:** Transient configuration-change defect, deterministic HTTP fixture only. No security analysis.

Source at audited revision: [backend/app/services/embeddings.py:132-136](https://github.com/ZaxbyHub/ragappv3/blob/0728825ad68fff81a016be2e4fc2f97a190d0621/backend/app/services/embeddings.py#L132)
```
        URL, model name, and prefixes are read live from ``settings`` on each
        call so admins can change them via the Settings UI without restarting.
        The initial URL is validated here to fail fast at startup when the
        service is unconfigured; subsequent reads tolerate an empty URL (the
        operation will fail at request time with a clear error).
```

### RERANK-001 — Local reranker ignores model changes after its first load.

**Disposition at planning:** Historical confirmed finding; re-anchor before implementation.

Changing the local reranker model leaves ranking on the old model until process restart. Separate instances with different explicit models also share the first model. Evidence: `backend/app/services/reranking.py:39-57`.

**Required repair direction:** Key the model cache by model identity, or atomically replace a single cached model when the configured identity changes; define lifetime and concurrency behavior.

**Regression that must fail before repair:** Call local scoring for two configured model identities and require each to invoke its corresponding model, with same-model calls still sharing a load.

**Evidence limit:** Local mode only; constructor fixture establishes cache identity, not model quality. No security assessment.

Source at audited revision: [backend/app/services/reranking.py:39-57](https://github.com/ZaxbyHub/ragappv3/blob/0728825ad68fff81a016be2e4fc2f97a190d0621/backend/app/services/reranking.py#L39)
```
def _get_local_model(model_id: str):
    """Lazy-load and cache a sentence-transformers CrossEncoder."""
    global _local_model
    if _local_model is None:
        with _model_lock:
            # Double-check after acquiring lock
            if _local_model is None:
                try:
                    from sentence_transformers import CrossEncoder
[excerpt continues in linked source]
```

### RERANK-002 — One feedback vote is multiplied by the number of passages from a document.

**Disposition at planning:** Historical confirmed finding; re-anchor before implementation.

Documents with more passages in an answer receive a larger positive or negative feedback adjustment from the same user rating, reaching the cap prematurely. Evidence: `backend/app/services/feedback_reranker.py:109-124`.

**Required repair direction:** Aggregate distinct message/document pairs before summing ratings.

**Regression that must fail before repair:** One rated answer with two passages from the same document must contribute one vote and one per-vote adjustment for that document.

**Evidence limit:** Functional count bias within one document collection only; no tenancy/security analysis. Default adjustment is capped at .10, limiting impact.

Source at audited revision: [backend/app/services/feedback_reranker.py:109-124](https://github.com/ZaxbyHub/ragappv3/blob/0728825ad68fff81a016be2e4fc2f97a190d0621/backend/app/services/feedback_reranker.py#L109)
```
            select_clause = """
                SELECT
                    cs.vault_id,
                    json_extract(src.value, '$.file_id') AS file_id,
                    SUM(CASE WHEN cm.feedback = 'up'   THEN 1 ELSE 0 END) AS up_votes,
                    SUM(CASE WHEN cm.feedback = 'down' THEN 1 ELSE 0 END) AS down_votes
                FROM chat_messages cm
[excerpt continues in linked source]
```

### RERANK-004 — TEI endpoint scores receive sigmoid normalization twice.

**Disposition at planning:** Historical confirmed finding; re-anchor before implementation.

TEI normalized scores from 0 to 1 become approximately .5 to .731. Even a near-zero response is displayed as Relevant. Ordering remains unchanged. Evidence: `backend/app/services/reranking.py:207-213`.

**Required repair direction:** Explicitly request raw_scores=True and normalize once, or retain normalized TEI scores directly.

**Regression that must fail before repair:** A normalized TEI response of .1 remains .1 under the chosen explicit protocol; raw logits are converted exactly once.

**Evidence limit:** Applies to TEI endpoints following the documented current protocol; local CrossEncoder behavior is not inferred. Primary-source references in artifacts/tei-rerank-contract.md.

Source at audited revision: [backend/app/services/reranking.py:207-213](https://github.com/ZaxbyHub/ragappv3/blob/0728825ad68fff81a016be2e4fc2f97a190d0621/backend/app/services/reranking.py#L207)
```
            "query": query,
            "texts": texts,
            "top_n": top_n,
            "truncate": True,
        }
        await asyncio.to_thread(assert_url_safe, self.reranker_url)
        if self._http_client is None:
```

## Supplemental audit and enhancement ownership

These named obligations supplement the individually numbered defect evidence. They are not additional independent defect counts. Current implementation requirements above and explicit acceptance criteria govern closure.

- [ ] **FULL-ENH-01** — Add an optional total prompt budget matched to the selected model — HIGH value, 0-30 days.

- [ ] **FULL-ENH-03** — Keep sentence deduplication from monopolizing the request event loop — MEDIUM value, 30-90 days.

- [ ] **FULL-ENH-04** — Move optional Redis cache I/O off the request event loop — MEDIUM value, 30-90 days.

- [ ] **FULL-ENH-05** — Validate legacy chunk-enrichment response shape before using auxiliary fields — MEDIUM value, 90+ days.

- [ ] **RAG-DEEP-01** — Successful decomposition discards embeddings already computed.

- [ ] **RAG-DEEP-02** — Decomposition repeats lexical work and reranking before deduplication.

- [ ] **RAG-DEEP-03** — Most document retrieval unnecessarily waits for wiki/KMS completion.

- [ ] **RAG-DEEP-04** — Existing context and serving costs compound the above.

## Legacy backlog obligations carried forward

Historical claims below are preserved for accountability, not accepted as current facts. Rebaseline stale absence/line-count claims before coding. The roadmap index splits cross-cutting bullets into accountable owners.

- [ ] **#229 legacy-01** — **Wire up the existing `EmbeddingSemanticChunker`** (`chunking.py:315`) — complete and unit-tested but read by zero call sites; live path is title-based fixed-size only. Highest-value low-effort win.

- [ ] **#229 legacy-03** — **Token-based context management.** History windowing is count-based (20 msgs, `prompt_builder.py:192-195`) and packing uses `len/3.5` char→token estimates; move to real tokenizer counts and detect `finish_reason` truncation mid-stream.

- [ ] **#229 legacy-04** — **Query decomposition** for multi-part questions (step-back/HyDE exist; decomposition doesn't).

## Semantic chunking and precise performance obligations

The semantic chunking strategy is already wired as an opt-in; old #229 absence wording is superseded. Verify its production call path rather than rebuilding it. Preserve optional status until F1/F2 quality and latency evidence supports broader rollout. Decomposition is already implemented. Reuse its computed embeddings; merge/deduplicate before shared lexical/reranking work; start independent document and Wiki/KMS retrieval concurrently while retaining source-specific budgets and deterministic final fusion. Include query-family and short-fact recall controls. E1 owns configurable thinking_max_tokens; B2 owns total tokenizer-aware prompt/output budgeting and consumes that setting.

## Shared rag_engine ownership

B1 coordinates the orchestration contract across B2, B3, E1 and E3, all of which edit rag_engine.py or its callers. B1 owns retrieval/control semantics; B2 owns call scheduling and context budgets; B3 owns post-retrieval vision; E1 owns maintenance, provider exception/circuit/fallback behavior (including OPS-005/006 and LLM-002/003); E3 owns stream metrics reset and admission hooks. Agree extraction points and tests before parallel edits, use additive interfaces, and review/rebase shared hunks with the affected owner. No slot silently rewrites another slot's behavior.

**Comment by zaxbysauce** [comment]
Implementation PR for this workstream B2 slot: **https://github.com/ZaxbyHub/ragappv3/pull/528** (head `48940c4f`, branch `fix/511-b2-model-aware-inference-budgets`).

**What changed** — all six owned findings repaired (EMBED-001 Ollama dialect parity incl. legacy per-item batch fan-out; EMBED-002 per-request config snapshots; RERANK-001 identity-keyed local model cache; RERANK-002 DISTINCT (message, document) vote aggregation; RERANK-003 honest singleton bypass; RERANK-004 TEI `raw_scores: true` single conversion) plus the supplemental obligations (FULL-ENH-01 optional total prompt budget with tiktoken-or-conservative token accounting and `finish_reason` capture; FULL-ENH-03 off-loop capped sentence dedup; FULL-ENH-04 bounded off-loop Redis cache I/O; FULL-ENH-05 per-field enrichment validation; RAG-DEEP-01 lazy variant embeddings; RAG-DEEP-02 consolidate-then-rerank — one rerank call per orchestration with facet F1 = 1.0; RAG-DEEP-03 KMS overlap with wiki precedence preserved). The two legacy-backlog claims were rebaselined as superseded per this issue's own wording (semantic chunker wired opt-in at `document_processor.py`; decomposition implemented in `query_transformer.py`).

**How to use** — no migration required. New settings (all runtime-settable via PUT /settings and restart-durable): `retrieval_consolidated_rerank` (True), `retrieval_kms_overlap` (True), `prompt_budget_enabled` (False — set `model_context_tokens` to your model's window before enabling), `prompt_reserve_output_tokens` (2048), `redis_io_timeout_seconds` (1.0), `context_distiller_max_sentences` (600). Rollback: the two scheduling flags disable at runtime; see the PR's release note (`docs/releases/pending/511-b2-redundant-work-model-budgets.md`).

**Evidence** — 17 frozen acceptance checks all RED→GREEN (independently re-run and mutation-probed by a separate reviewer context; 4.6 final critic APPROVE after one docs-only revision), 230 new regression test cases, impacted existing suites green, SAST baseline flat at 137 (zero new patterns). Full trace: `.agents/issue-traces/511-b2-model-aware-inference-budgets/` on the branch.
